### ISSUE TITLE
[raysgd] Allow for fractional GPU resources in RaySGD

### ISSUE THREAD

**Issue by DavidMChan** [ray-project/ray#9258]
### Describe your feature request

Currently, RaySGD allocates a single GPU if "use_gpu" is enabled, however with the updates to Ray's scheduling (schedules fractionally based on GPU memory usage), it can be impossible to schedule workers on GPUs that are idle with the exception of an X-server, or a background process.

I believe this is a relatively simple change to https://github.com/ray-project/ray/blob/e0d8b58969728054d1800abc925aa5352882baa8/python/ray/util/sgd/torch/torch_trainer.py#L333 and https://github.com/ray-project/ray/blob/e0d8b58969728054d1800abc925aa5352882baa8/python/ray/util/sgd/torch/torch_trainer.py#L339, however it's not entirely clear.

If it is as simple as introducing a new argument "num_gpus_per_worker", and propagating the value, I can submit a PR for this.

**Comment by richardliaw** [comment]
Hey David, thanks for opening this issue! What changes to ray scheduling are you referring to re: utilization?

**Comment by DavidMChan** [comment]
I have been running into issues where a job won't schedule a 1 GPU job with 0.8 GPUs available - so I assumed that there had been changes to the scheduling to account for GPUs being occupied by non-ray processes. If this isn't correct, then I'll have to take another look at my code. 

Regardless, this is probably a good feature to have for non-memory limited jobs in RaySGD.

**Comment by richardliaw** [comment]
Got it - is "0.8 GPUs" quantity determined by Ray? or is it some other non-ray process? I think Ray doesn't take into account the utilization from other non-ray processes. 

BTW, the problem with fractional GPU resources is that you'll want each worker to be placed on a separate GPU, while using less than one GPU resource. This works for a quantity between 0.5 and 1, but if your resource quantity is 0.25, it's likely that all models get placed on the same GPU. Unfortunately, Ray doesn't provide great control for this right now..

**Comment by DavidMChan** [comment]
Yep - the 0.8 is a quantity from Ray:

```
2020-07-01 14:47:53,858 WARNING worker.py:1116 -- The actor or task with ID ffffffffffffffff6f53dca10100 is pending and cannot currently be scheduled. It requires {CPU: 1.000000}, {GPU: 1.0000
00} for execution and {CPU: 1.000000}, {GPU: 1.000000} for placement, but this node only has remaining {node:128.32.175.26: 0.800000}, {CPU: 10.000000}, {memory: 35.400391 GiB}, {object_store_
memory: 12.207031 GiB}. In total there are 0 pending tasks and 2 pending actors on this node. This is likely due to all cluster resources being claimed by actors. To resolve the issue, conside
r creating fewer actors or increase the resources available to this Ray cluster. You can ignore this message if this Ray cluster is expected to auto-scale.
```

Nothing else is running in Ray, so it's odd that something is taking 0.2 GPUs.

**Comment by richardliaw** [comment]
oh, 0.8 is just for bookkeeping (reserving cpus); it isn’t related to the GPU usage. What’s odd is you don’t seem to have any GPUs detected?

**Comment by richardliaw** [comment]
RE: model ensembles; yeah that sounds interesting; I still don't have a great idea of what to do about that. I guess ideally, you'll just define it all in your model creator.

**Comment by DavidMChan** [comment]
Hmm, seems like another issue then - I'll go ahead and close this for the time being.

**Comment by DavidMChan** [comment]
For future reference - I'm having the same issue as in #8326 - not really sure what's going on though.

**Comment by richardliaw** [comment]
Are GPUs not being detected? Can you open a new issue and provide compute environment details? 

Would love to get you unblocked asap!
